### PR TITLE
[build] Added options to make tags using github API

### PR DESCRIPTION
**Pull request by majisourav99**
To push and release tags on github we need committer info. This PR adds support to pass github token to make API calls to github for adding tags.

Ran the workflow on fork and also did curl call to create the tag locally.

### CODE REVIEW

File: make_tag.py [hunk 1]
```diff
@@ -19,12 +20,13 @@
 @click.option('--bump-major', is_flag=True)
 @click.option('--bump-minor', is_flag=True)
 @click.option('--no-verify', is_flag=True)
-def read_config(bump_major, bump_minor, no_verify):
+@click.option('--github-token', default = ' ', help='The github token used for API call')
```
Comment (nisargthakkar) [inline make_tag.py:23]: Let's make the default as `None`?

File: make_tag.py [hunk 2]
```diff
@@ -127,20 +129,23 @@ def make_tag(remote, bump_major, bump_minor, need_verification):
             print('Skipped creating the tag')
             return
 
-    tag_success = call(['git', 'tag', '-a', '-m', tag_message, tag_name])
-    if tag_success != 0:
-        sys.exit()
-    call(['git', 'push', remote, tag_name])
+    if github_token != ' ':
+        os.system('./curltag.sh {} {}' .format(str(tag_name), str(github_token)))
```
Comment (nisargthakkar) [inline make_tag.py:133]: Let's use `requests` to make the API calls instead of curl and shell script

File: curltag.sh [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+repo=$(basename $remote .git)
+# POST a new ref to repo via Github API
+curl -s -X POST https://api.github.com/repos/linkedin/venice/$repo/git/refs \
+    -H "Authorization: token $2" \
+    -d @- << EOF
+    {
+       "ref": "refs/tags/$1",
+       "sha": "$commit"
+    }
+    EOF
```
Comment (nisargthakkar) [inline curltag.sh:15]: Can you add some test section to the PR description?

**Review by nisargthakkar — COMMENTED** [review]
Please edit PR title to be more descriptive and add a `[build]` tag

File: .github/workflows/build-and-upload-archives-on-schedule.yml [hunk 4]
```diff
@@ -2,7 +2,7 @@ name: Venice Publication Pipeline
 
 on:
   schedule:
-    - cron: "30 0 * * 3"
+    - cron: "30 20 * * 3"
   workflow_dispatch:
 
 jobs:
```
Comment (nisargthakkar) [inline .github/workflows/build-and-upload-archives-on-schedule.yml:8]: If the deploy keys work, we'll need to pass them via env vars. Check this for an example in another action: https://github.com/linkedin/venice/blob/main/.github/workflows/build-and-upload-archives-upon-creating-tags.yml#L38

**Review by nisargthakkar — CHANGES_REQUESTED** [review]

File: make_tag.py [hunk 5]
```diff
@@ -127,20 +129,30 @@ def make_tag(remote, bump_major, bump_minor, need_verification):
...
-    tag_success = call(['git', 'tag', '-a', '-m', tag_message, tag_name])
-    if tag_success != 0:
-        sys.exit()
-    call(['git', 'push', remote, tag_name])
+    if github_token:
+        headers = {
+            'Authorization': f'token {github_token}',
+            'Content-Type': 'application/x-www-form-urlencoded',
+        }
+        response = requests.post('https://api.github.com/repos/linkedin/venice/git/refs', headers=headers, json={"ref": "refs/tags/" + tag_name, "sha" : "$commit"})
```
Comment (nisargthakkar) [inline make_tag.py:137]: `$commit` needs to be an actual commit reference

two nits:
1. Let's use `''` where possible (to follow the pattern of the rest of the file)
2. we can use `fstring` for `tag_name` now: `f'refs/tags/{tag_name}'`

File: make_tag.py [hunk 6]
```diff
@@ -127,20 +129,30 @@ def make_tag(remote, bump_major, bump_minor, need_verification):
...
+        if (response.status_code != 200):
+            print("Could not push the tag " + tag_name)
```
Comment (nisargthakkar) [inline make_tag.py:139]: Let's change `push` to `create`
Also, same nit - use `''`

File: make_tag.py [hunk 7]
```diff
@@ -19,12 +20,13 @@
 @click.option('--bump-major', is_flag=True)
 @click.option('--bump-minor', is_flag=True)
 @click.option('--no-verify', is_flag=True)
-def read_config(bump_major, bump_minor, no_verify):
+@click.option('--github-token',  help='The github token used for API call')
```
Comment (nisargthakkar) [inline make_tag.py:23]: nit: extra whitespace before `help`

File: make_tag.py [hunk 8]
```diff
@@ -127,20 +129,32 @@ def make_tag(remote, bump_major, bump_minor, need_verification):
...
-    tag_success = call(['git', 'tag', '-a', '-m', tag_message, tag_name])
-    if tag_success != 0:
-        sys.exit()
-    call(['git', 'push', remote, tag_name])
+    if github_token:
+        headers = {
+            'Authorization': f'token {github_token}',
+            'Content-Type': 'application/x-www-form-urlencoded',
+        }
+        commit = call(['git', 'rev-parse', 'HEAD'])
```
Comment (nisargthakkar) [inline make_tag.py:137]: The return value of `call` is the command's exit status

**Review by nisargthakkar — CHANGES_REQUESTED** [review]

File: make_tag.py [hunk 9]
```diff
@@ -127,20 +129,32 @@ def make_tag(remote, bump_major, bump_minor, need_verification):
...
             return
 
-    tag_success = call(['git', 'tag', '-a', '-m', tag_message, tag_name])
-    if tag_success != 0:
-        sys.exit()
-    call(['git', 'push', remote, tag_name])
+    if github_token:
+        headers = {
+            'Authorization': f'token {github_token}',
+            'Content-Type': 'application/x-www-form-urlencoded',
```
[the remark this answers is not in the document]
Comment (majisourav99) [reply make_tag.py:135]: you mean the header?

[the remark this answers is not in the document]
Comment (nisargthakkar) [reply make_tag.py:135]: The `Content-Type` specifically

[the remark this answers is not in the document]
Comment (majisourav99) [reply make_tag.py:135]: some SO article I saw used application/json. But guess it not necessary

[the remark this answers is not in the document]
Comment (nisargthakkar) [reply make_tag.py]: If we use `json=...` in `requests`, then the library adds the header automatically

File: curltag.sh (hunk 3 above)
Comment (nisargthakkar) [reply curltag.sh]: @majisourav99 Can you add this?

**Review by nisargthakkar — APPROVED** [review]
LGTM

**Pull request merged**